### ISSUE TITLE
Register adaptors on module install

### ISSUE THREAD

**Issue by jvegasbsc** [radical-cybertools/saga-python#459]
Hi

I have created an adaptors for a machine that have a "special" PBS installation. This adaptor is supposed to be distributed with the main application package (autosubmit). I don´t know which is the preferred method to register an adaptor for SAGA from another package. 

Thanks in advance for your help!

Javier

**Comment by andre-merzky** [comment]
Hi Javier, 

SAGA has a static list of adaptors, see https://github.com/radical-cybertools/saga-python/blob/devel/src/saga/engine/registry.py .  Including adaptors from other packages would probably need a more dynamic approach.  Let me ponder over the options for a bit, will ping back.

**Comment by andre-merzky** [comment]
Hi again, 

would you mind checking the `feature/external_adaptors` branch?  There you should be able to do something like the following:

```
$ unset SAGA_ADAPTOR_PATH 
$ python -c 'import saga; s=saga.Session()' 2>&1 | grep loading |tail -n 3
2015-09-01 12:30:18,809: radical.saga        : MainProcess                     : MainThread     : INFO    : loading  adaptor saga.adaptors.globus_online.go_file
2015-09-01 12:30:18,810: radical.saga        : MainProcess                     : MainThread     : INFO    : loading  adaptor saga.adaptors.torque.torquejob
2015-09-01 12:30:18,812: radical.saga        : MainProcess                     : MainThread     : INFO    : loading  adaptor saga.adaptors.pbspro.pbsprojob

$ export SAGA_ADAPTOR_PATH=saga.adaptors.shell.shell_job:saga.adaptors.shell.shell_file
$ python -c 'import saga; s=saga.Session()' 2>&1 | grep loading |tail -n 5
2015-09-01 12:30:22,898: radical.saga        : MainProcess                     : MainThread     : INFO    : loading  adaptor saga.adaptors.globus_online.go_file
2015-09-01 12:30:22,900: radical.saga        : MainProcess                     : MainThread     : INFO    : loading  adaptor saga.adaptors.torque.torquejob
2015-09-01 12:30:22,903: radical.saga        : MainProcess                     : MainThread     : INFO    : loading  adaptor saga.adaptors.pbspro.pbsprojob
2015-09-01 12:30:22,904: radical.saga        : MainProcess                     : MainThread     : INFO    : loading  adaptor saga.adaptors.shell.shell_job
2015-09-01 12:30:22,906: radical.saga        : MainProcess                     : MainThread     : INFO    : loading  adaptor saga.adaptors.shell.shell_file
```

In other words, you should be able to point `SAGA_ADAPTOR_PATH` to the location of module install for your adaptor, and it should then be picked up at that location.  Is that approximately what you were looking for?

**Comment by jvegasbsc** [comment]
Hi

Your quickness is awesome! I think this will be enough for my needs. I think that you add all adaptors in the SAGA_ADAPTOR_PATH to what is already on the registry.py, I'm correct? I can not check out this branch today, but I will try it later this week.

Thanks!

Javier

**Comment by andre-merzky** [comment]
We are not always this quick I'm afraid :P  But this was something which we actually have planned for in the past, so its not much code change, really...

You are correct: the content of `SAGA_ADAPTOR_PATH` is dynamically added to the registry, and then the normal adaptor loading kicks in.  Take your time with testing :)

**Comment by jvegasbsc** [comment]
I have installed the branch in a virtualenv and get this error
```
[CRITICAL] Unhandled exception on Autosubmit: 'module' object has no attribute 'get_logger'
Traceback (most recent call last):
  File "/home/jvegas/pyCharm/autosubmit/autosubmit/autosubmit.py", line 253, in parse_args
    return Autosubmit.run_experiment(args.expid)
  File "/home/jvegas/pyCharm/autosubmit/autosubmit/autosubmit.py", line 508, in run_experiment
    submitter.load_platforms(as_conf)
  File "/home/jvegas/pyCharm/autosubmit/autosubmit/platforms/submitter.py", line 41, in load_platforms
    local_platform.service = saga.job.Service("fork://localhost")
  File "/home/jvegas/virtualenvs/autosubmit/local/lib/python2.7/site-packages/saga/job/service.py", line 109, in __init__
    session = ss.Session (default=True)
  File "/home/jvegas/virtualenvs/autosubmit/local/lib/python2.7/site-packages/saga/session.py", line 222, in __init__
    simple_base.__init__ ()
  File "/home/jvegas/virtualenvs/autosubmit/local/lib/python2.7/site-packages/saga/base.py", line 36, in __init__
    self._logger = ru.get_logger  ('radical.saga')
AttributeError: 'module' object has no attribute 'get_logger'
```

I have not yet modify anything on the code to use the new feature. It is working fine with current SAGA version

**Comment by marksantcroos** [comment]
You will also need the development version of radical.utils now.
You can find that here: https://github.com/radical-cybertools/radical.utils

**Comment by jvegasbsc** [comment]
Everything it's working fine now.

Thanks!

**Comment by jvegasbsc** [comment]
I'm getting the same error that issue #428. I don know if it is supposed to be corrected on this branch.
```
[CRITICAL] Unhandled exception on Autosubmit: stop waiting on object lease
Traceback (most recent call last):
  File "/home/jvegas/pyCharm/autosubmit/autosubmit/autosubmit.py", line 253, in parse_args
    return Autosubmit.run_experiment(args.expid)
  File "/home/jvegas/pyCharm/autosubmit/autosubmit/autosubmit.py", line 574, in run_experiment
    Autosubmit.submit_ready_jobs(as_conf, joblist, platforms_to_test)
  File "/home/jvegas/pyCharm/autosubmit/autosubmit/autosubmit.py", line 615, in submit_ready_jobs
    platform.send_file(scriptname)
  File "/home/jvegas/pyCharm/autosubmit/autosubmit/platforms/platform.py", line 118, in send_file
    out = saga.filesystem.File("file://{0}".format(os.path.join(self.tmp_path, filename)))
  File "/home/jvegas/virtualenvs/autosubmit/local/lib/python2.7/site-packages/saga/filesystem/file.py", line 86, in __init__
    _adaptor, _adaptor_state, _ttype=_ttype)
  File "/home/jvegas/virtualenvs/autosubmit/local/lib/python2.7/site-packages/saga/namespace/entry.py", line 89, in __init__
    url, flags, session, ttype=_ttype)
  File "/home/jvegas/virtualenvs/autosubmit/local/lib/python2.7/site-packages/saga/base.py", line 101, in __init__
    self._init_task = self._adaptor.init_instance (adaptor_state, *args, **kwargs)
  File "/home/jvegas/virtualenvs/autosubmit/local/lib/python2.7/site-packages/saga/adaptors/cpi/decorators.py", line 57, in wrap_function
    return sync_function (self, *args, **kwargs)
  File "/home/jvegas/virtualenvs/autosubmit/local/lib/python2.7/site-packages/saga/adaptors/shell/shell_file.py", line 1009, in init_instance
    self.local = self.lm.lease (lease_tgt, self.shell_creator, lease_tgt)
  File "/home/jvegas/virtualenvs/autosubmit/local/lib/python2.7/site-packages/radical/utils/lease_manager.py", line 371, in lease
    raise LookupError ("stop waiting on object lease")
LookupError: stop waiting on object lease
```

**Comment by andre-merzky** [comment]
I am closing this ticket -- the object lease problem should be unrelated to the adaptor loading.  Please open a new ticket if this persists!
